python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py

```diff
@@ -450,13 +451,35 @@ def foo(x, y):
                 out_foo = decorated_foo(paddle.rand([10]), paddle.rand([10]))
                 print(decorated_foo.concrete_program)
         """
+        return self.concrete_program_specify_input_spec(input_spec=None)
+
+    def concrete_program_specify_input_spec(self, input_spec=None):
+        """
+        Returns recent ConcreteProgram instance of decorated function while
+        specifying input_spec. If the self._function_spec already has
+        input_spce, it will check the compatibility of input input_spec and
+        the self._function_spec.input_spec. If input input_spec=None, then
+        this method uses self._function_spec.input_spec
+
+        args:
+            input_spec (list[InputSpec], optional): Describes the input of
+                the translate function.
+        """
         # if specific the `input_spec`, the length of program_cache will always 1,
         # else, return the last one.
         cached_program_len = len(self._program_cache)
         # If specific `input_spec`, apply convertion from dygraph layers into static Program.
         if cached_program_len == 0:
-            input_spec = self._function_spec.input_spec
-            has_input_spec = (input_spec is not None and len(input_spec) > 0)
+            if input_spec is None:
+                input_spec = self._function_spec.input_spec
+            elif self._function_spec.input_spec is not None:
+                if not input_specs_compatible(input_spec,
+                                              self._function_spec.input_spec):
```

Comment:
Should we add `flatten(input_spec)` and `flatten(sellf._function_spec.input_spec)` here to make sure all them are pure list? Because we support nested input spec structure.

Reply:
Good suggestion, done!